liota/dccs/iotcv2.py

```diff
@@ -0,0 +1,346 @@
+                        if json_msg["result"] == 'SUCCESS':
+                            self.set_relationship = True
+                    else:
+                        log.info("Waiting for create relationship response")
+                        on_response(rel_resp_q.get(True, timeout), rel_resp_q)
+                except:
+                    raise Exception("Exception while create relationship")
+
+            # create and add register request into req_list (waiting list)
+            operation_id = self.next_id()
+            req = Request(reg_entity_parent.reg_entity_id, operation_id, rel_resp_q)
+            log.info("operation_id:{0} req:{1}".format(operation_id, req))
+            with self.req_ops_lock:
+                self.req_dict.update({str(operation_id): req})
+                log.info("operation_id:{0} req_dict:{1}".format(operation_id, self.req_dict))
+            self.comms.send(json.dumps(self._relationship(operation_id,
+                            reg_entity_parent.reg_entity_id, reg_entity_child.reg_entity_id)))
+            # block until there is an item available or timeout
+            try:
+                on_response(rel_resp_q.get(True, timeout), rel_resp_q)
+            except:
+                raise Exception("Exception while creating relationship")
+
+            if not self.set_relationship:
+                raise CreateRelationshipFailure()
+            log.info("Relationship Created")
+
+    def _format_data(self, reg_metric):
+        met_cnt = reg_metric.values.qsize()
+        if met_cnt == 0:
+            return
+        data = []
+        for _ in range(met_cnt):
+            m = reg_metric.values.get(block=True)
+            if m is not None:
+                data.append({"ts": m[0], "v": m[1]})
+        if len(data) == 0:
+            return
+
+        metrics = []
+        metrics_entry = {
+            "metricName": reg_metric.ref_entity.name,
+            "data": data
+        }
+        metrics.append(metrics_entry)
+
+        # TBD: now, only 1 metric_data_entry, and for 1 MO
+        metric_data = []
+        metric_data_entry = {
+            "MOguid": reg_metric.parent.ref_entity.entity_id,
+            "metrics": metrics
+        }
+        metric_data.append(metric_data_entry)
+        return json.dumps({
+            "type": "putMetric_req",
+            "metric_data": metric_data
+        })
+
+    def set_properties(self, reg_entity, properties):
+        # RegisteredMetric get parent's guid; RegisteredEntity gets own guid
+
+        if isinstance(reg_entity, RegisteredMetric):
+            entity = reg_entity.parent.ref_entity
+        else:
+            entity = reg_entity.ref_entity
+
+        log.info("Properties defined for resource {0}".format(entity.name))
+        # set_properties fun internal queue
+        set_prop_resp_q = Queue.Queue()
+        self.set_prop = False
+
+        def on_response(msg, set_prop_resp_q):
```

Comment:
If we are doing the same kind of check everywhere, why can't we extract this out as a single function and call it with parameters everywhere? Why are we duplicating the code? Am i missing something here?

Reply:
They are not the same, but have similar procedure. They have different if condition check. Do not want to mess them up with too many input arguments. The on_response codes only have ~ 10 lines.